#6654, #9974: make include and with constraints aware of ghost components

Octachron opened the pull request:
Classes, class types and private row types add ghost components to the signature where they are defined.

For instance,
```ocaml
module type s = sig
  class c: object end
end
```
is expanded in the typed signature to
```ocaml
module type s = sig
    class c: object end
    class type c= object end
    type c
    type #c
end
```

When editing or printing a signature it is therefore important to identify those ghost components.

Before this PR, the logic to recognize ghost components was reimplemented (or forgotten) in few different places.
This PR (in collaboration with @gasche) proposes to centralize this logic in a new `Signature_group` group module.
This module provides signature iterators that works of group of core component and associated ghost components. 

Once defined, those iterators have be reused to deal in a unified way with ghost components in printing, shadowing included components and editing signature with `with_constraint`.

## Printing (e3af864)

For `Printtyp`, this PR is merely using the new iterator which are a slightly less specialized version of the pre-existing code.

## Shadowing included components (206fbfa) 

Using structured iterators make it easier to remove all components of a (real+ghost) group:
```ocaml
 include sig class c: object end end type c
```
now results with the `type c` shadowing the whole class `c`, ghost class type `c` etc... rather than silently erasing the ghost type `c` of the class.
In other words, the following code type checks:
```ocaml
module Ok: sig
  include sig class c: object end end type c
end =
struct type c end
```
We could be less lenient and raises an error however.
This is implemented by remembering the full group of core and ghost components for shadowable items, and removing all of them simultaneously when needed.
  
## Editing signature through a `with` constraint (9ce0b9b , c52f1d0)  

Inversely, once ghost components are identified, we can make sure that  `with` constraint cannot edit them.
For instance, trying to steal the ghost type of a component
```ocaml
  module type a = sig class type ct = object end end with type ct := < >
``` 
raises an error
>```
> Error: The signature constrained by `with' has no component named ct
>```
with this PR rather than an `assert false` in `Subst`.
This is kind of solving #6654 by keeping more consistently ghost component out of sight of users.

The implementation simply move the `editing in place` logic of `Typemod.merge_constraint` to `Signature_group` and let
`Typemod.merge_constraint` focuses on the type system.

Fix #9974 (really this time!)

gasche committed:
[refactoring] Typemod.simplify: use List.filter_map for readability

Co-authored-by: Florian Angeletti <florian.angeletti@inria.fr>

Octachron committed:
Signature_group: ghost-aware iteration over signatures

Signatures contain ghost type items that depends on a core syntactic
item, the module "Signature_group" provides iterators that are aware
of this difference between syntactic and ghost items.

Octachron committed:
include shadow all items in a group:

When shadowing a componrnt which is either a ghost component
or a syntactic component with associated ghost components, we
remove both the syntactic component and the related ghosts.

This fix issues in cases like

  include struct class c = object end end
  type c = int

where only the ghost type c from the class definition was shadowed, thus
breaking the class definition.

The new behavior removes the whole class definition keeping the
class definition format.

Octachron committed:
with constraint: ignore ghost components

"With" constraint should not be able to manipulate
independently ghost signature components. For instance, in

  module type s = sig class c: object end end with type c := < >

the compiler should fail with a missing component item rather than
silently remove the ghost type of the class c.

Octachron committed:
Signature_group: replace_in_place

Octachron committed:
update Changes

Octachron committed:
more documentation

gasche commented on typing/signature_group.mli:
```diff
@@ -0,0 +1,43 @@
+val seq: Types.signature -> rec_group Seq.t
+
+val iter: (rec_group -> unit) -> Types.signature -> unit
+val fold: ('acc -> rec_group -> 'acc) -> 'acc -> Types.signature -> 'acc
```
Why do we need to re-export `iter` and `fold`? I would just provide `seq` and compose with `Seq` operators, instead of duplicating a fragment of the Seq API that will keep growing as we want more stuff. But maybe you have the opposite aesthetic preference and I can be convinced?

Octachron replied:
They both survived a change of representation (from a `fold` to a `sequence`), and it seemed simpler to keep them while the representation is still partially in flux.

gasche commented on typing/signature_group.ml:
```diff
@@ -0,0 +1,116 @@
+
+let rec partial_lists l () = match l with
+  | [] -> Seq.Nil
+  | a :: q -> Seq.Cons((a,q), partial_lists q)
```
I don't like this `partial_lists` business, it feels too complex (and not so composable). Can you explain when the apparently-redundant `q` data is important/useful?

Octachron replied:
It is used to edit only one item in a signature in the case of a `with` constraint without traversing the whole signature.
The other option would be to reconstruct the remaining signature from the structured sequence.
Or I could be also convinced to switch to a zipper.

gasche commented:
I would be curious to see a version of this patch with the following interface:

```ocaml
module Signature_group : sig
  type rec_group

  [...]

  val append : rec_group -> Types.signature_item list -> Types.signature_item list
  val rev_append : rec_group -> Types.signature_item list -> Tyeps.signature_item list  (* actually those are Revlist.t *)

  val next : Types.signature_item list -> (rec_group * Types.signature_item list) option
  (* raises Invalid_arg if 'take' fails *)

  val seq : Types.signature_item list -> rec_group Seq.t
  let seq = Seq.unfold next_group
end
```

and `replace_in_place` specialized at the callsite, following the general pattern

```ocaml
let rec replace sg acc =
  match Signature_group.next sg with
  | None -> List.rev acc
  | Some (group, rem) ->
  let continue () = replace (Signature_group.rev_append group acc) rem in
  let return group' = List.rev_append acc (Signature_group.append group' rem) in
  ...
```

I think that this may strike a good balance:
- the types and concepts involved are simpler than in the current iteration (and yes, slightly more low-level at places)
- if we later decide that we care about the performance of `merge_constraint`, we can implement a version that tries to fuse the replace of several constraints in one traversal, without having to change the interface of `Signature_group`.

I wanted to prototype this myself but I didn't get the time so far, so I'm writing it down in case someone else wants to -- or has good feedback for why that would not turn out to be an improvement.

Octachron commented:
I am not really convinced since your proposal is transferring back the complexity of the iteration inside the recursive group to `Typemod.signature_constraint` and would rather transfer complexity away from `Typemod`.

gasche commented:
A compromise would be to remove `full_seq` and `to_partials`, define `seq` as `Seq.unfold next`, and implement `replace` in terms of `next` directly instead of going through your clever but weird accumulation sequence.

Octachron commented:
Going through `unfold` sounds like a sensible compromise if you prefer to get rid `full_seq`.

gasche commented:
I'm reviewing the "user" commits (the shadowing one is fine). I would suggest waiting to see if there are more suggestions coming out of that before you put work into changing the current patchset.

gasche commented on typing/signature_group.mli:
```diff
@@ -0,0 +1,85 @@
+*)
+val replace_in_place:
+  ( rec_group:sig_item list -> ghosts:Types.signature -> Types.signature_item
+    -> 'a in_place_patch option )
```
This could be `('a * in_place_patch) option`.

gasche replied:
(I still think that using `('a * in_place_patch) option` would simplify the code slightly, by making `in_place_patch` simpler and highlighting the regularity between the return type of the user function and the return type of the whole functional.)

gasche commented on typing/typemod.ml:
```diff
@@ -557,33 +560,36 @@ let merge_constraint initial_env loc sg lid constr =
-       (With_type sdecl | With_typesubst sdecl as constr))
+        return ~ghosts
+          ~replace_by:[ Sig_type(id, newdecl, rs, priv);
+                        Sig_type(id_row, decl_row, rs', priv)
```
If I understand correctly, here the second `Sig_type` is morally a ghost type declaration, but we don't say it explicitly because we are expanding into a non-grouped list anyway. Is that right? Would the code look nicer if we tried to reconstruct a valid group first, and then flatten it within the `replace` function?

Octachron replied:
In fact, we need to transfer to it the ghost group to fix a bug in trunk. Currently
```ocaml
module type s = sig
  type t
  and u = private < .. >
  and v
end with type u = private < .. >
```
is printed
```ocaml
module type s = sig type t and u#row and u = private < .. > and v end
```
because the `with` constraint breaks the assumption that all ghost row components are prefixed to the group. 
I have an incoming patch that does this.

gasche commented:
I'm done reviewing the code. Thoughts:

- This is nice work. (I participated to the original design discussion and a small bit of hacking for the small first commit, but not the rest.)
- The two changes to the actual type-checking logic (the shadowing stuff and the translation of `with` constraints) are nice. They are also not very invasive so it's reassuring that they are less likely to introduce bugs.
- I am unsure about the implementation of the `Signature_group` which I find too complex. I am not sure how important that is: maybe we should just take in the current implementation and work on improving the type-checker further, instead of optimizing it for partly-subjective design goals.

In the medium term, do we still want to move to a representation of signatures that contains the the grouping, instead of re-computing this structure on the fly? If yes, maybe we should move to that, and bet that the details we are quibbling over will just disappear in the move.

Octachron committed:
bugfix: with constraints and #row component

This fix makes `with` constraints preserves the location of
ghost #row component:

   sig type t and v = private < .. > and u end with type v = private <..>

no longer moves the location of the v#row component to just before v.

With this change, we are enforcing the invariants that ghost row
components always precede the mutually recursive group of definitions.

gasche commented:
With the bugfix, `replace_by` is now a list of at most one element. Questions:
- Should this be an `option` instead?
- Does it make sense to have at least one ghost item if the replacement patch removes the "main" item? If I understand correctly, the answer is currently "yes", because the ghost items of the whole recursive group as passed together (`split_row_id` is precisely in charge of locating the one ghost item associated to that main item, if it exists).

gasche commented on typing/typemod.ml:
```diff
@@ -511,12 +499,23 @@ let merge_constraint initial_env loc sg lid constr =
+    split [] ghosts
+  in
+  let rec patch_item constr namelist sig_env ~rec_group ~ghosts item =
+    let return ?(ghosts=ghosts) ?(replace_by=[]) info =
```
I'm not fond of having `replace_by` be an optional argument, it makes it more difficult to locate the items that get removed by the transformation.

Octachron commented:
Replacing the list of new elements by an option is probably a good idea.

Concerning the ghost part, yes, `merge_constraint` is doing the editing of the ghost item for the whole group. On that point, the sharing of responsibility feels suboptimal. I will think more about it.

If we are planning to switch to a more structured signature on an intermediate scale, I personally think that it is not necessary to spend too much time optimizing the on-the-fly iterators. Conversely, starting with only iterators first gives us some room to iterate over the structured signature design without committing visible changes for users yet.

gasche reviewed (approved):
I agree with the general assessment, and on that basis I approve the PR.

gasche commented on Changes:
```diff
@@ -382,6 +382,16 @@ Working version
+   the type `c` shadows the `class c` to avoid shadowing only the ghost type
+   c introduced by the class.
+  (Florian Angeletti and Gabriel Scherer, report by Eduardo Rafael,
+   review by ???)
```
I would propose to move me away from the authors and into the reviewers. (We discussed the general direction beforehand but the code is all yours, except for the first trivial refactoring commit.)

Octachron committed:
signature_group review: next + unfold

Octachron committed:
review: Changes

Octachron committed:
review: with_constraints delete or replace one item

gasche reviewed (approved):
Thanks for the version with `next`, this now looks good, and I'm in favor of merging.

Octachron committed:
signature_group review: more symmetric interface

Octachron commented:
I was finally convinced the symmetry argument for the `in_place_patch` type.

gasche commented:
Maybe I should also have pointed out earlier that "some more data to return to the outer scope" is bad enough documentation that it should suggest to kill the corresponding record field :-}
